pkg/api/latest/latest.go

```diff
@@ -18,131 +18,108 @@ package latest
 
 import (
 	"fmt"
-	"sort"
-	"strings"
 
+	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/meta"
 	"k8s.io/kubernetes/pkg/api/registered"
 	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
-var (
-	allGroups = GroupMetaMap{}
-	// Group is a shortcut to allGroups.Group.
-	Group = allGroups.Group
-	// RegisterGroup is a shortcut to allGroups.RegisterGroup.
-	RegisterGroup = allGroups.RegisterGroup
-	// GroupOrDie is a shortcut to allGroups.GroupOrDie.
-	GroupOrDie = allGroups.GroupOrDie
-	// AllPreferredGroupVersions returns the preferred versions of all
-	// registered groups in the form of "group1/version1,group2/version2,..."
-	AllPreferredGroupVersions = allGroups.AllPreferredGroupVersions
-	// IsRegistered is a shortcut to allGroups.IsRegistered.
-	IsRegistered = allGroups.IsRegistered
-)
+// ExternalVersions lists all allowed external versions for this group from most preferred to least preferred
+var ExternalVersions = []unversioned.GroupVersion{}
 
-// ExternalVersions is a list of all external versions for this API group in order of
-// most preferred to least preferred
-var ExternalVersions = []unversioned.GroupVersion{
-	{Group: "", Version: "v1"},
-}
+// PreferredExternalVersion is the most preferred external version (ExternalVersions[0])
+var PreferredExternalVersion = unversioned.GroupVersion{}
```

Comment:
This is not a concept I want in the code. No code should be talking about a preferred external version. (I see this is actually a GroupVersion, or I'd be complaining about the name and how versions don't make sense without groups.) No group is special, so there's no sense of "preferred group", either-- there won't be multiple groups that have the thing you want, and if there were, you still wouldn't have a *global* preferred group.

I maintain that code either
a) Knows exactly what group/version it needs in which case it should never reference this
b) Knows a group it needs, in which case it should not reference this
c) Can work with anything, in which case it should not use this variable.